lib/src/merged_tree.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A lazily merged view of a set of trees.
+
+use std::cmp::max;
+
+use itertools::Itertools;
+
+use crate::backend::TreeValue;
+use crate::conflicts::Conflict;
+use crate::repo_path::{RepoPath, RepoPathComponent};
+use crate::store::Store;
+use crate::tree::Tree;
+use crate::tree_builder::TreeBuilder;
+
+/// Presents a view of a merged set of trees.
+pub enum MergedTree {
+    /// A single tree, possibly with path-level conflicts.
+    Legacy(Tree),
+    /// A merge of multiple trees, or just a single tree. The individual trees
+    /// have no path-level conflicts.
+    Merge(Conflict<Tree>),
+}
+
+/// The value at a given path in a `MergedTree`.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub enum MergedTreeValue<'a> {
+    /// A single non-conflicted value.
+    Resolved(Option<&'a TreeValue>),
+    /// TODO: Make this a `Conflict<Option<&'a TreeValue>>` (reference to the
+    /// value) once we have removed the `MergedTree::Legacy` variant.
+    Conflict(Conflict<Option<TreeValue>>),
+}
+
+impl MergedTree {
+    /// Creates a new `MergedTree` representing a single tree without conflicts.
+    pub fn resolved(tree: Tree) -> Self {
+        MergedTree::new(Conflict::resolved(tree))
+    }
+
+    /// Creates a new `MergedTree` representing a merge of a set of trees. The
+    /// individual trees must not have any conflicts.
+    pub fn new(conflict: Conflict<Tree>) -> Self {
+        debug_assert!(!conflict.removes().iter().any(|t| t.has_conflict()));
+        debug_assert!(!conflict.adds().iter().any(|t| t.has_conflict()));
+        debug_assert!(itertools::chain(conflict.removes(), conflict.adds())
+            .map(|tree| tree.dir())
+            .all_equal());
+        debug_assert!(itertools::chain(conflict.removes(), conflict.adds())
+            .map(|tree| tree.store().as_ref() as *const Store)
```

Comment:
nit: `Arc::as_ptr(tree.store())` can be used

Reply:
Nice.